stress: failed test in cockroach/kv/kv.test: TestTxnCoordIdempotentCleanup

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/d6c1c5a04bfbd46b7fee8b02ecbad01b103c4034

Stress build found a failed test:

```
=== RUN   TestTxnCoordIdempotentCleanup
W160416 04:11:50.040985 gossip/gossip.go:888  not connected to cluster; use --join to specify a connected node
I160416 04:11:50.041019 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
I160416 04:11:50.042556 storage/replica_command.go:1430  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 1s
E160416 04:11:50.042696 config/config.go:382  unable to determine largest reserved object ID from system config: descriptor table not found in system config of 0 values
--- FAIL: TestTxnCoordIdempotentCleanup (5.02s)
	txn_coord_sender_test.go:767: kv/txn_coord_sender.go:473: writing transaction timed out or ran on multiple coordinators
	leaktest.go:82: Leaked goroutine: goroutine 1373 [chan receive]:
		github.com/cockroachdb/cockroach/rpc.NewContext.func1()
			/go/src/github.com/cockroachdb/cockroach/rpc/context.go:89 +0x57
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82058c2a0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1374 [chan receive]:
		github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc82062e660)
			/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:164 +0x3a
		created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
			/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:165 +0x8f4
	leaktest.go:82: Leaked goroutine: goroutine 1375 [select]:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc82062f3e0)
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:239 +0xd21
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:224 +0x20
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc820520440)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1376 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/store.go:1878 +0xbee
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc820520ca0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1377 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).Start.func3()
			/go/src/github.com/cockroachdb/cockroach/storage/store.go:794 +0x13d
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82058c9e0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1549 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc820574600)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1550 [select]:
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).waitAndProcess(0xc820248d20, 0xecea3b286, 0x28793ac, 0x2531060, 0xc820656100, 0xc820248230, 0xc82013a460, 0xc82038fe98)
			/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:162 +0x568
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1.1(0xc82013a460, 0xc8204f28a0)
			/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:204 +0x78
		github.com/cockroachdb/cockroach/storage.(*storeRangeSet).Visit(0xc8204f2ba0, 0xc82022f000)
			/go/src/github.com/cockroachdb/cockroach/storage/store.go:237 +0x246
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:206 +0x21e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc820574620)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1563 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076fe00)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1564 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076fe60)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1565 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076e000)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1566 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076e020)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1567 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076e080)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1568 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/queue.go:356 +0x303
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076e0a0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1569 [select]:
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).waitAndProcess(0xc820248850, 0xecea3b286, 0x287d765, 0x2531060, 0xc820656100, 0xc820248230, 0xc82013a460, 0xc82038e698)
			/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:162 +0x568
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1.1(0xc82013a460, 0xc8204f28a0)
			/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:204 +0x78
		github.com/cockroachdb/cockroach/storage.(*storeRangeSet).Visit(0xc8204f29f0, 0xc8206c2000)
			/go/src/github.com/cockroachdb/cockroach/storage/store.go:237 +0x246
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:206 +0x21e
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82076e0c0)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1570 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func1()
			/go/src/github.com/cockroachdb/cockroach/storage/store.go:865 +0x3cf
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82058ca00)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
	leaktest.go:82: Leaked goroutine: goroutine 1571 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func2()
			/go/src/github.com/cockroachdb/cockroach/storage/store.go:884 +0x3cf
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820248230, 0xc82058ca20)
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
0 runs so far, 0 failures, over 30s
2 runs so far, 0 failures, over 35s
5 runs so far, 0 failures, over 40s
6 runs so far, 0 failures, over 45s
7 runs so far, 0 failures, over 50s
8 runs so far, 0 failures, over 55s
8 runs so far, 0 failures, over 1m0s
9 runs so far, 0 failures, over 1m5s
9 runs so far, 0 failures, over 1m10s
9 runs so far, 0 failures, over 1m15s
12 runs so far, 0 failures, over 1m20s
14 runs so far, 0 failures, over 1m25s
16 runs so far, 0 failures, over 1m30s
16 runs so far, 0 failures, over 1m35s
17 runs so far, 0 failures, over 1m40s
17 runs so far, 0 failures, over 1m45s
18 runs so far, 0 failures, over 1m50s
20 runs so far, 0 failures, over 1m55s
21 runs so far, 0 failures, over 2m0s
23 runs so far, 0 failures, over 2m5s
24 runs so far, 0 failures, over 2m10s
24 runs so far, 0 failures, over 2m15s
25 runs so far, 0 failures, over 2m20s
25 runs so far, 0 failures, over 2m25s
27 runs so far, 0 failures, over 2m30s
29 runs so far, 0 failures, over 2m35s
29 runs so far, 0 failures, over 2m40s
30 runs so far, 0 failures, over 2m45s
32 runs so far, 0 failures, over 2m50s
33 runs so far, 0 failures, over 2m55s
33 runs so far, 0 failures, over 3m0s
34 runs so far, 0 failures, over 3m5s
35 runs so far, 0 failures, over 3m10s
36 runs so far, 0 failures, over 3m15s
37 runs so far, 0 failures, over 3m20s
38 runs so far, 0 failures, over 3m25s
39 runs so far, 0 failures, over 3m30s
40 runs so far, 0 failures, over 3m35s
41 runs so far, 0 failures, over 3m40s
41 runs so far, 0 failures, over 3m45s
43 runs so far, 0 failures, over 3m50s
45 runs so far, 0 failures, over 3m55s
46 runs so far, 0 failures, over 4m0s
46 runs so far, 0 failures, over 4m5s
47 runs so far, 0 failures, over 4m10s
47 runs so far, 0 failures, over 4m15s
49 runs so far, 0 failures, over 4m20s
50 runs so far, 0 failures, over 4m25s
51 runs so far, 0 failures, over 4m30s
53 runs so far, 0 failures, over 4m35s
54 runs so far, 0 failures, over 4m40s
54 runs so far, 0 failures, over 4m45s
56 runs so far, 0 failures, over 4m50s
56 runs so far, 0 failures, over 4m55s
57 runs so far, 0 failures, over 5m0s
58 runs so far, 0 failures, over 5m5s
60 runs so far, 0 failures, over 5m10s
60 runs so far, 0 failures, over 5m15s
60 runs so far, 0 failures, over 5m20s
62 runs so far, 0 failures, over 5m25s
62 runs so far, 0 failures, over 5m30s
62 runs so far, 0 failures, over 5m35s
64 runs so far, 0 failures, over 5m40s
65 runs so far, 0 failures, over 5m45s
66 runs so far, 0 failures, over 5m50s
68 runs so far, 0 failures, over 5m55s
68 runs so far, 0 failures, over 6m0s
69 runs so far, 0 failures, over 6m5s
69 runs so far, 0 failures, over 6m10s
69 runs so far, 0 failures, over 6m15s
71 runs so far, 0 failures, over 6m20s
72 runs so far, 0 failures, over 6m25s
74 runs so far, 0 failures, over 6m30s
75 runs so far, 0 failures, over 6m35s
75 runs so far, 0 failures, over 6m40s
76 runs so far, 0 failures, over 6m45s
77 runs so far, 0 failures, over 6m50s
78 runs so far, 0 failures, over 6m55s
78 runs so far, 0 failures, over 7m0s
79 runs so far, 0 failures, over 7m5s
81 runs so far, 0 failures, over 7m10s
83 runs so far, 0 failures, over 7m15s
83 runs so far, 0 failures, over 7m20s
84 runs so far, 0 failures, over 7m25s
85 runs so far, 0 failures, over 7m30s
85 runs so far, 0 failures, over 7m35s
87 runs so far, 0 failures, over 7m40s
87 runs so far, 0 failures, over 7m45s
87 runs so far, 0 failures, over 7m50s
91 runs so far, 0 failures, over 7m55s
92 runs so far, 0 failures, over 8m0s
92 runs so far, 0 failures, over 8m5s
93 runs completed, 1 failures, over 8m8s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/kv/txn_coord_sender.go
+++ b/kv/txn_coord_sender.go
@@ -18,6 +18,7 @@
 package kv
 
 import (
+	"errors"
 	"fmt"
 	"sync"
 	"sync/atomic"
@@ -45,6 +46,9 @@ const (
 	opTxnCoordSender  = "txn coordinator"
 	opHeartbeatLoop   = "heartbeat"
 )
+
+var errNoState = errors.New("writing transaction timed out " +
+	"or ran on multiple coordinators")
 
 // txnMetadata holds information about an ongoing transaction, as
 // seen from the perspective of this coordinator. It records all
@@ -328,64 +332,65 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*r
 		if err := tc.maybeBeginTxn(&ba); err != nil {
 			return nil, roachpb.NewError(err)
 		}
-		txnID := *ba.Txn.ID
-		if pErr := tc.maybeRejectClient(ctx, *ba.Txn); pErr != nil {
+		var et *roachpb.EndTransactionRequest
+		var hasET bool
+		{
+			var rArgs roachpb.Request
+			rArgs, hasET = ba.GetArg(roachpb.EndTransaction)
+			if hasET {
+				et = rArgs.(*roachpb.EndTransactionRequest)
+				if len(et.Key) != 0 {
+					return nil, roachpb.NewErrorf("EndTransaction must not have a Key set")
+				}
+				et.Key = ba.Txn.Key
+				if len(et.IntentSpans) > 0 {
+					// TODO(tschottdorf): it may be useful to allow this later.
+					// That would be part of a possible plan to allow txns which
+					// write on multiple coordinators.
+					return nil, roachpb.NewErrorf("client must not pass intents to EndTransaction")
+				}
+			}
+		}
+
+		if pErr := func() *roachpb.Error {
+			tc.Lock()
+			defer tc.Unlock()
+			if pErr := tc.maybeRejectClientLocked(ctx, *ba.Txn); pErr != nil {
+				return pErr
+			}
+
+			if !hasET {
+				return nil
+			}
+			// Everything below is carried out only when trying to commit.
+
+			// Populate et.IntentSpans, taking into account both any existing
+			// and new writes, and taking care to perform proper deduplication.
+			var keys interval.RangeGroup
+			if txnMeta, metaOK := tc.txns[*ba.Txn.ID]; metaOK {
+				keys = txnMeta.keys
+			} else {
+				keys = interval.NewRangeTree()
+			}
+			ba.IntentSpanIterate(func(key, endKey roachpb.Key) {
+				addKeyRange(keys, key, endKey)
+			})
+			et.IntentSpans = collectIntentSpans(keys)
+			if len(et.IntentSpans) == 0 {
+				// If there aren't any intents, then there's factually no
+				// transaction to end. Read-only txns have all of their state
+				// in the client.
+				return roachpb.NewErrorf("cannot commit a read-only transaction")
+			}
+			return nil
+		}(); pErr != nil {
 			return nil, pErr
 		}
 
-		if rArgs, ok := ba.GetArg(roachpb.EndTransaction); ok {
-			et := rArgs.(*roachpb.EndTransactionRequest)
-			if len(et.Key) != 0 {
-				return nil, roachpb.NewErrorf("EndTransaction must not have a Key set")
-			}
-			et.Key = ba.Txn.Key
-			if len(et.IntentSpans) > 0 {
-				// TODO(tschottdorf): it may be useful to allow this later.
-				// That would be part of a possible plan to allow txns which
-				// write on multiple coordinators.
-				return nil, roachpb.NewErrorf("client must not pass intents to EndTransaction")
-			}
-			tc.Lock()
-			txnMeta, metaOK := tc.txns[txnID]
-			{
-				// Populate et.IntentSpans, taking into account both existing
-				// writes (if any) and new writes in this batch, and taking
-				// care to perform proper deduplication.
-				var keys interval.RangeGroup
-				if metaOK {
-					keys = txnMeta.keys
-				} else {
-					keys = interval.NewRangeTree()
-				}
-				ba.IntentSpanIterate(func(key, endKey roachpb.Key) {
-					addKeyRange(keys, key, endKey)
-				})
-				et.IntentSpans = collectIntentSpans(keys)
-			}
-			tc.Unlock()
-
-			if len(et.IntentSpans) > 0 {
-				// All good, proceed.
-			} else if !metaOK {
-				// If we don't have the transaction, then this must be a retry
-				// by the client. We can no longer reconstruct a correct
-				// request so we must fail.
-				//
-				// TODO(bdarnell): if we had a GetTransactionStatus API then
-				// we could lookup the transaction and return either nil or
-				// TransactionAbortedError instead of this ambivalent error.
-				return nil, roachpb.NewErrorf("transaction is already committed or aborted")
-			}
-			if len(et.IntentSpans) == 0 {
-				// If there aren't any intents, then there's factually no
-				// transaction to end. Read-only txns have all of their state in
-				// the client.
-				return nil, roachpb.NewErrorf("cannot commit a read-only transaction")
-			}
-			if log.V(1) {
-				for _, intent := range et.IntentSpans {
-					log.Trace(ctx, fmt.Sprintf("intent: [%s,%s)", intent.Key, intent.EndKey))
-				}
+		if hasET && log.V(1) {
+			for _, intent := range et.IntentSpans {
+				log.Trace(ctx, fmt.Sprintf("intent: [%s,%s)",
+					intent.Key, intent.EndKey))
 			}
 		}
 	}
@@ -444,10 +449,10 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*r
 	return br, nil
 }
 
-// maybeRejectClient checks whether the (transactional) request is in a state
-// that prevents it from continuing, such as the coordinator having considered
-// the client abandoned, or a heartbeat having reported an error.
-func (tc *TxnCoordSender) maybeRejectClient(
+// maybeRejectClientLocked checks whether the (transactional) request is in a
+// state that prevents it from continuing, such as the coordinator having
+// considered the client abandoned, or a heartbeat having reported an error.
+func (tc *TxnCoordSender) maybeRejectClientLocked(
 	ctx context.Context,
 	txn roachpb.Transaction,
 ) *roachpb.Error {
@@ -455,8 +460,6 @@ func (tc *TxnCoordSender) maybeRejectClient(
 	if !txn.Writing {
 		return nil
 	}
-	tc.Lock()
-	defer tc.Unlock()
 	txnMeta, ok := tc.txns[*txn.ID]
 	// Check whether the transaction is still tracked and has a chance of
 	// completing. It's possible that the coordinator learns about the
@@ -469,8 +472,7 @@ func (tc *TxnCoordSender) maybeRejectClient(
 		// transaction session so that we can definitively return the right
 		// error between these possible errors. Or update the code to make an
 		// educated guess based on the incoming transaction timestamp.
-		return roachpb.NewErrorf("writing transaction timed out " +
-			"or ran on multiple coordinators")
+		return roachpb.NewError(errNoState)
 	case txnMeta.txn.Status == roachpb.ABORTED:
 		txn := txnMeta.txn.Clone()
 		tc.cleanupTxnLocked(ctx, txn)

--- a/kv/txn_coord_sender_test.go
+++ b/kv/txn_coord_sender_test.go
@@ -742,7 +742,7 @@ func TestTxnCoordSenderTxnUpdatedOnError(t *testing.T) {
 	}
 }
 
-// TestTxnCoordIdempotentCleanup verifies that cleanupTxn is idempotent.
+// TestTxnCoordIdempotentCleanup verifies that cleanupTxnLocked is idempotent.
 func TestTxnCoordIdempotentCleanup(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	s := createTestDB(t)
@@ -757,13 +757,18 @@ func TestTxnCoordIdempotentCleanup(t *testing.T) {
 	}
 
 	s.Sender.Lock()
+	// Clean up twice successively.
+	s.Sender.cleanupTxnLocked(context.Background(), txn.Proto)
 	s.Sender.cleanupTxnLocked(context.Background(), txn.Proto)
 	s.Sender.Unlock()
 
+	// For good measure, try to commit (which cleans up once more if it
+	// succeeds, which it may not if the previous cleanup has already
+	// terminated the heartbeat goroutine)
 	ba = txn.NewBatch()
 	ba.InternalAddRequest(&roachpb.EndTransactionRequest{})
 	pErr := txn.Run(ba)
-	if pErr != nil && !testutils.IsPError(pErr, "aborted") {
+	if pErr != nil && !testutils.IsPError(pErr, errNoState.Error()) {
 		t.Fatal(pErr)
 	}
 }

